src/main/java/minijava/lexer/Lexer.java

```diff
@@ -409,34 +406,24 @@ private Token createToken(Terminal terminal, String content) {
     return new Token(terminal, position, content);
   }
 
-  private Token nextToken() {
-    if (current != null && current.isEOF()) {
-      return current;
-    }
-    if (lookAheadBuffer.size() > 0) {
-      current = lookAheadBuffer.get(0);
-      lookAheadBuffer.remove(0);
-    } else {
-      current = parseNextToken();
-    }
-    return current;
-  }
-
   @Override
   public boolean hasNext() {
-    return numberOfEOFs < 1;
+    return token == null || !token.isEOF();
```

Comment:
`token == null` is only true once (in the beginning). _**If**_ we were optimizing for performance, the test for `!token.isEOF()` should probably be the first one :stuck_out_tongue: